qep-13-efficient-snapping-and-geometry-queries.rst

```diff
@@ -0,0 +1,234 @@
+as a core functionality.
+
+Ideally the implementation should focus just on search for the closest vertices/segments
+and leave extra features for higher level - to keep the querying core simple and make it easier to optimize it.
+
+The aim is to simplify the implementation to the following structure:
+
+1. **low-level querying** - for a set of geometries (typically one vector layer), find out the nearest
+   vertex/segment or all near vertices/segments within tolerance.
+2. **high-level querying** - operates on top of several layers, combines results from lower level
+3. **extra functionality** - routines that are not directly related to querying, but are useful in context of snapping
+   and are provided for user's convenience
+
+Currently the snapping uses either simple geometry cache or iterates over layers if cache is not available.
+The cache however does not have an efficient data structure for spatial querying (data are stored
+in a map with feature IDs as keys and geometries as values). Introduction of R-trees should improve
+the performance of queries even with greater amount of geometries and greater complexity of those (from
+linear search to logarithmic). This should enable more interactive GUI functionality later,
+such as mouse over effects for various tools or continuous mode for identify tool
+(like Value Tool, not having to click on each feature).
+
+Implementation Details
+----------------------
+
+1. new class ``QgsPointLocator`` [core lib]
+
+   - operates on one vector layer, watches the layer for changes (added/changed/removed features)
+   - defines ``Match`` class for retrieving results
+   - builds data structures for efficient point location queries (R trees)
+   - supports queries for matches within tolerance (vertex/edge), k-nearest neighbors (vertex/edge), point in polygon
+   - supports reprojection (i.e. data may be indexed in different CRS)
+   - supports lazy indexing (unless needed for a request, data are not indexes)
+   - supports filtering of matches (e.g. accept matches only from a particular feature ID) with ``MatchFilter``
+   - replacement for ``QgsSnapper``
+
+
+2. new class ``QgsSnappingUtils`` [core lib]
+
+   - keeps a cache of QgsPointLocator classes
+   - snapToMap method: combine results from several QgsPointLocator instances according to setup
+   - two modes for snapToMap
+   
+       1. snap to current layer (with default snapping type and tolerance)
+       2. snap to all layers from configuration (advanced)
+   - not directly connected to map canvas - need map settings and current layer to be explicitly set
+   - optional snapping on intersections
+   - client code can access locators
+   - supports filtering of matches (e.g. accept matches only from a particular feature ID) with ``MatchFilter``
+   - configuration can be read from project (but does not need to be)
+   - replacement for ``QgsMapCanvasSnapper``
+
```

Comment:
Yes it should be easy to implement such strategy. It could be a new "ordered" mode where the search would be stopped once first good match is found. I thought about using this strategy by default for QgsSnappingUtils, but the project's snapping configuration is not an ordered list so...